pkg/store/limiter.go

```diff
@@ -82,16 +86,89 @@ func NewChunksLimiterFactory(limit uint64) ChunksLimiterFactory {
 	}
 }
 
-// NewSeriesLimiterFactory makes a new NewSeriesLimiterFactory with a static limit.
+// NewSeriesLimiterFactory makes a new SeriesLimiterFactory with a static limit.
 func NewSeriesLimiterFactory(limit uint64) SeriesLimiterFactory {
 	return func(failedCounter prometheus.Counter) SeriesLimiter {
 		return NewLimiter(limit, failedCounter)
 	}
 }
 
-// NewSeriesLimiterFactory makes a new NewSeriesLimiterFactory with a static limit.
+// NewBytesLimiterFactory makes a new BytesLimiterFactory with a static limit.
 func NewBytesLimiterFactory(limit units.Base2Bytes) BytesLimiterFactory {
 	return func(failedCounter prometheus.Counter) BytesLimiter {
 		return NewLimiter(uint64(limit), failedCounter)
 	}
 }
+
+type RateLimits struct {
+	SeriesPerRequest  uint64
+	SamplesPerRequest uint64
+}
+
+func (l *RateLimits) RegisterFlags(cmd extkingpin.FlagClause) {
+	cmd.Flag("store.grpc.series-limit", "The maximum series allowed for a single Series request. The Series call fails if this limit is exceeded. 0 means no limit.").Default("0").Uint64Var(&l.SeriesPerRequest)
+	cmd.Flag("store.grpc.samples-limit", "The maximum samples allowed for a single Series request, The Series call fails if this limit is exceeded. 0 means no limit. NOTE: For efficiency the limit is internally implemented as 'chunks limit' considering each chunk contains a maximum of 120 samples.").Default("0").Uint64Var(&l.SamplesPerRequest)
+}
+
+// rateLimitedStoreServer is a storepb.StoreServer that can apply rate limits against Series requests.
+type rateLimitedStoreServer struct {
+	storepb.StoreServer
+	newSeriesLimiter      SeriesLimiterFactory
+	newSamplesLimiter     ChunksLimiterFactory
+	failedRequestsCounter *prometheus.CounterVec
+}
+
+// NewRateLimitedStoreServer creates a new rateLimitedStoreServer.
+func NewRateLimitedStoreServer(store storepb.StoreServer, reg prometheus.Registerer, rateLimits RateLimits) storepb.StoreServer {
+	return &rateLimitedStoreServer{
+		StoreServer:       store,
+		newSeriesLimiter:  NewSeriesLimiterFactory(rateLimits.SeriesPerRequest),
+		newSamplesLimiter: NewChunksLimiterFactory(rateLimits.SamplesPerRequest),
+		failedRequestsCounter: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_selects_dropped_total",
+			Help: "Number of select queries that were dropped due to configured limits.",
+		}, []string{"reason"}),
+	}
+}
+
+func (s *rateLimitedStoreServer) Series(req *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
+	seriesLimiter := s.newSeriesLimiter(s.failedRequestsCounter.WithLabelValues("series"))
+	chunksLimiter := s.newSamplesLimiter(s.failedRequestsCounter.WithLabelValues("chunks"))
+	rateLimitedSrv := newRateLimitedServer(srv, seriesLimiter, chunksLimiter)
+	if err := s.StoreServer.Series(req, rateLimitedSrv); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// rateLimitedServer is a storepb.Store_SeriesServer that tracks statistics about sent series.
+type rateLimitedServer struct {
```

Comment:
Should we add the classic assignment to interface type to ensure this type keeps implementing the `storepb.Store_SeriesServer` interface at compile time?